consensus/recovery/recover.go

```diff
@@ -1,60 +1,108 @@
+	return nil
+}
+
+// ForksState recovers Forks' internal state of blocks descending from the latest
+// finalized block. Caution, input blocks must be valid and in parent-first order
+// (unless parent is the latest finalized block).
+func ForksState(forks hotstuff.Forks) BlockScanner {
+	return func(proposal *model.Proposal) error {
+		err := forks.AddValidatedBlock(proposal.Block)
+		if err != nil {
+			return fmt.Errorf("could not add block %v to forks: %w", proposal.Block.BlockID, err)
 		}
-		if errors.Is(err, model.ErrUnverifiableBlock) {
-			log.Warn().
-				Hex("block_id", logging.ID(proposal.Block.BlockID)).
-				Hex("qc_block_id", logging.ID(proposal.Block.QC.BlockID)).
-				Msg("unverifiable proposal")
-
-			// even if the block is unverifiable because the QC has been
-			// pruned, it still needs to be added to the forks, otherwise,
-			// a new block with a QC to this block will fail to be added
-			// to forks and crash the event loop.
-		} else if err != nil {
-			return fmt.Errorf("cannot validate proposal (%x): %w", proposal.Block.BlockID, err)
+		return nil
+	}
+}
+
+// VoteAggregatorState recovers the VoteAggregator's internal state as follows:
+//   - Add all blocks descending from the latest finalized block to accept votes.
+//     Those blocks should be rapidly pruned. As the node catches up.
+//
+// Caution: input blocks must be valid.
+func VoteAggregatorState(voteAggregator hotstuff.VoteAggregator) BlockScanner {
+	return func(proposal *model.Proposal) error {
+		voteAggregator.AddBlock(proposal)
+		return nil
+	}
+}
+
+// CollectParentQCs collects all parent QCs included in the blocks descending from the
+// latest finalized block. Caution, input blocks must be valid.
+func CollectParentQCs(collector Collector[*flow.QuorumCertificate]) BlockScanner {
+	return func(proposal *model.Proposal) error {
+		qc := proposal.Block.QC
+		if qc != nil {
+			collector.Append(qc)
 		}
+		return nil
+	}
+}
 
-		err = onProposal(proposal)
-		if err != nil {
-			return fmt.Errorf("cannot recover proposal: %w", err)
+// CollectTCs collect all TCs included in the blocks descending from the
+// latest finalized block. Caution, input blocks must be valid.
+func CollectTCs(collector Collector[*flow.TimeoutCertificate]) BlockScanner {
+	return func(proposal *model.Proposal) error {
+		tc := proposal.LastViewTC
+		if tc != nil {
+			collector.Append(tc)
 		}
+		return nil
 	}
+}
 
-	log.Info().Msgf("recovery completed")
+// Collector for objects of generic type. Essentially, it is a stateful list.
+// Safe to be passed by value. Retrieve() returns the current state of the list
+// and is unaffected by subsequent appends.
+type Collector[T any] struct {
+	list *[]T
+}
 
-	return nil
+func NewCollector[T any]() Collector[T] {
+	list := make([]T, 0, 5) // heuristic: pre-allocate with some basic capacity
+	return Collector[T]{list: &list}
 }
+
+func (c Collector[T]) Append(t ...T) { *c.list = append(*c.list, t...) }
```

Comment:
Very elegant, it took me a minute to understand why `Retrieve` is not effected by subsequent appends.

Reply:
Thanks for the comment. You needing to stop and think why `Retrieve` is not effected by subsequent appends indicates that I should add a comment with a brief explanation. Chances are high, I'll get stuck at exactly the same point when I look at my own code in 6 months 😅.

Added this: 
https://github.com/onflow/flow-go/blob/134fa8b4a4063ff8fdc08ccb26b1096edd9a2d06/consensus/recovery/recover.go#L112-L119

Reply:
:star_struck: